File: src/year2020/month11/no5/LongestPalindromicSubstring.java
package year2020.month11.no5;

public class LongestPalindromicSubstring {
    public static void main(String[] args) {
        String s1 = "babad";    //bab, aba
        String s2 = "cbbd";     //bb
        String s3 = "a";        //a
        String s4 = "ac";       //a, c
        String s5 = "aaaaaaa";       //aaaaaaa
        System.out.println(longestPalindrome3(s1));
        System.out.println(longestPalindrome3(s2));
        System.out.println(longestPalindrome3(s3));
        System.out.println(longestPalindrome3(s4));
        System.out.println(longestPalindrome3(s5));
    }

    private static String longestPalindrome3(String s) {
//        dynamic programming
        int begin = 0;
        int maxLen = 0;
        boolean[][] dp = new boolean[s.length()][s.length()];
//        len means the length of s[i...j]
        for (int len = 1; len <= s.length(); len++) {
            for (int i = 0; i <= s.length() - len; i++) {
                int j = i + len - 1;
                if (len == 1) {
                    dp[i][j] = true;
                } else if (len == 2) {
                    dp[i][j] = s.charAt(i) == s.charAt(j);
                } else {
                    dp[i][j] = s.charAt(i) == s.charAt(j) && dp[i + 1][j - 1];
                }
                if (dp[i][j] && maxLen < len) {
                    maxLen = len;
                    begin = i;
                }
            }
        }
        return s.substring(begin, begin + maxLen);
    }

    private static String longestPalindrome2(String s) {
        if (s.length() < 2) {
            return s;
        }
//        the longest palindrome shoule be the s[begin,end)
        int begin = 0;
        int end = 0;
        char[] chars = s.toCharArray();
        for (int i = 0; i < chars.length; i++) {
//            consider the palindrome is odd or even, and compare to get the one that more longer
            int oddLen = centerExpand(chars, i, i);
            int evenLen = centerExpand(chars, i, i + 1);
            int len = Math.max(oddLen, evenLen);
            if (len > end - begin) {
                begin = i - (len - 1) / 2;
                end = begin + len;
            }
        }
        return s.substring(begin, end);
    }

    private static int centerExpand(char[] chars, int left, int right) {
        while (left >= 0 && right < chars.length && chars[left] == chars[right]) {
            left--;
            right++;
        }
        return right - left - 1;
    }

    private static String longestPalindrome1(String s) {
//        certer expand, add separator # to make sub palindrome string is odd
        int begin = 0;
        int len = 0;
        String tempString = getTempString(s);
        for (int i = 0; i < tempString.length(); i++) {
            int left = i;
            int right = i;
            while (left >= 0 && right < tempString.length() && tempString.charAt(left) == tempString.charAt(right)) {
                left--;
                right++;
            }
            if (len < right - left - 1) {
                len = right - left - 1;
                begin = left + 1;
            }
        }
        return tempString.substring(begin, begin + len).replace("#", "");
    }

    private static String getTempString(String s) {
        StringBuilder sb = new StringBuilder("#");
        for (int i = 0; i < s.length(); i++) {
            sb.append(s.charAt(i)).append("#");
        }
        return new String(sb);
    }
}

/*
* 给定一个字符串 s，找到 s 中最长的回文子串。你可以假设 s 的最大长度为 1000。

示例 1：

输入: "babad"
输出: "bab"
注意: "aba" 也是一个有效答案。
示例 2：

输入: "cbbd"
输出: "bb"

来源：力扣（LeetCode）
链接：https://leetcode-cn.com/problems/longest-palindromic-substring
著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
* */

File: go/year2025/month5/no5/LongestPalindromicSubstring.go
package main

func main() {
	s1 := "babad"
	s2 := "cbbd"
	println(longestPalindrome(s1)) // "bab" or "aba"
	println(longestPalindrome(s2)) // "bb"
}

// 中心拓展法，times: O(n^2)，space: O(1)
// 枚举中心点，向两边扩展
func longestPalindrome(s string) string {
	start := 0
	end := 0
	for i := 0; i < len(s); i++ {
		// 以s[i]为中心的回文串
		len1 := expandAroundCenter(s, i, i)
		// 以s[i]和s[i+1]为中心的回文串
		len2 := expandAroundCenter(s, i, i+1)
		maxLen := max(len1, len2)
		if maxLen > end-start {
			start = i - (maxLen-1)/2 // 考虑奇偶数
			end = start + maxLen - 1
		}
	}
	return s[start : end+1]
}

func expandAroundCenter(s string, left int, right int) int {
	for left >= 0 && right < len(s) && s[left] == s[right] {
		left--
		right++
	}
	// 此时left和right都不匹配或者越界了
	return right - left - 1
}

/*
DP, times: O(n^2), space: O(n^2)
状态定义：dp[i][j]表示s[i:j]是否是回文串
状态转移方程：dp[i][j] = dp[i+1][j-1] && s[i] == s[j], i < j，只需要填上半表格
转移方向：左下到右上，len: 2 -> len(s), i: 0 -> len(s), j = len - i - 1
初始化：dp[i][i] = true
*/
func longestPalindrome1(s string) string {
	dp := make([][]bool, len(s))
	for i := 0; i < len(dp); i++ {
		dp[i] = make([]bool, len(s))
		dp[i][i] = true // 单个字符是回文串
	}
	longestStart := 0
	longestEnd := 0
	for l := 2; l <= len(s); l++ { // 长度从小往大拓展
		for i, j := 0, l-1; j < len(s); i, j = i+1, j+1 { // l = j - i + 1
			if l >= 3 {
				dp[i][j] = dp[i+1][j-1] && s[i] == s[j]
			} else {
				dp[i][j] = s[i] == s[j]
			}
			if dp[i][j] && j-i > longestEnd-longestStart { // 更新最长回文子串
				longestStart = i
				longestEnd = j
			}
		}
	}
	return s[longestStart : longestEnd+1]
}

/*
给你一个字符串 s，找到 s 中最长的 回文 子串。

示例 1：
输入：s = "babad"
输出："bab"
解释："aba" 同样是符合题意的答案。

示例 2：
输入：s = "cbbd"
输出："bb"

提示：
1 <= s.length <= 1000
s 仅由数字和英文字母组成
*/
